### ISSUE TITLE
disable whichwrap 

### ISSUE THREAD

**Issue by irevoire** [mawww/kakoune#3057]
Hello!
I discovered Kakoune some hours ago and I must say I really like the idea. But after one day there is an option I don’t find.

In vim by default when you are at the start of a line and press the left arrow or h your cursor stay at the column 0.
Same thing for the end of the line when you go to the right.

Then with the [whichwrap](https://vim.fandom.com/wiki/Automatically_wrap_left_and_right) option you can configure this feature.

Since I really hate my cursor going everywhere when I’m working on a line I was wondering if there is a plan to implements this feature?
Or even better, if I missed it?

Thanks in advance,
Best.

**Comment by lenormf** [comment]
The editor has actually moved away from keeping the cursor onto the current line to allowing it to wrap around, so I don't think it will be enforced again in the foreseeable future.

As far as implementing that feature, I guess some custom hooks (on `NormalKey`) that check for the position of anchor/cursor would work.

**Comment by lenormf** [comment]
You could try the following:

```
hook global NormalKey h %{
    try %{
        execute-keys <a-k>$<ret> l
    }
}

hook global NormalKey l %{
    try %{
        execute-keys <a-k>^<ret>h
    }
}
```

There's just one bug: when you have a multiple selection, the one that has its cursor already on the very first character of the buffer will disappear upon hitting `h`. Not sure if it's expected or not.

**Comment by irevoire** [comment]
Oh thanks a lot I was wondering how I could do it!
For now it’ll be ok to continue my learning 👍 

I was thinking maybe this should be in the migrating from vim page.

**Comment by lenormf** [comment]
Feel free to document what works for you in the wiki, anybody can edit it!

**Comment by irevoire** [comment]
Hello @lenormf I was wondering, how would you do to also remove the whichwrap option when deleting text?
For example when in this position with `|` being the cursor:
```
bonjour.|
I love cheese
```
And you press `d`, coming from vim you expect the `.` to be deleted instead of the `\n`.

I guess the hook is stil on `NormalKey` but this time with the letter `d`. But I have no idea on which command you could run to check if you are at the end of the line and choose to delete backward instead of forward

**Comment by lenormf** [comment]
There is always a selection in a buffer, the cursor cannot be between two letters, it is at the very least "over" one character at all times.

In your example, the cursor is over the newline character (i.e. the current selection contains the newline character).

You hit the <kbd>d</kbd> key, which by default erases the current selection(s). So the following line gets appended to the current one.

I'm not sure it's worth it to force the line-by-line model using hooks, the editing experience remains good without it (if not better, arguably), but if you still want that then you can adapt the code I posted earlier.

**Comment by lenormf** [comment]
Turns out the implementation isn't that trivial, so here it is:

```
define-command delete-whichwrap %{ evaluate-commands -draft -itersel %sh{
    if [ "${kak_selection}" = "
" ]; then
        echo "exec i<backspace><esc>"
    else
        echo "exec d"
    fi
} }
map global normal d ': delete-whichwrap<ret>'
```

**Comment by irevoire** [comment]
Oh ok I think I got it, the thing is in vim the last character of a line is not the new line:
![image](https://user-images.githubusercontent.com/7032172/63163680-6c504180-c026-11e9-940f-fc43e2cc8dac.png)
Here I can't go more on the right for example even if there is a newline after the `h`.
So I think in vim you just can't select the new line.

Whatever thanks for the implementation I would not be able to do it alone.
I still find quite odd to not provide an option for this as in vim / emacs, but I think with this setting I'll be able to use the editor without having too much trouble :+1:

**Comment by lenormf** [comment]
I found out why my original code was buggy when a selection was at the beginning of the buffer, could you please try out this snippet for <kbd>h</kbd>?

```
hook global NormalKey h %{
    try %{
        # If the cursor is not at the top of the buffer already
        execute-keys -draft <a-H> K <a-K>\A^[^\n]*\z<ret>
        # And it is now located on the previous line's newline character.
        # Move it back to the right
        execute-keys <a-k>$<ret> l
    }
}
```

If it works as expected, we'll replace the code in the wiki and I'll come up with a corresponding hook for <kbd>l</kbd>.

**Comment by irevoire** [comment]
Hello, sorry for the delay I was out for the week end.
I just tried the code, it works but I'm not sure I see a difference with the previous snippet :frowning: 
I may have missed the original bug?

![image](https://user-images.githubusercontent.com/7032172/63248756-5762f080-c268-11e9-9ee8-c1de4a64a3cc.png)

With one cursor when I try to do a selection to the left I got this message "no selections remaining".
And with multiple cursor, the cursor already on the left disappear with the two snippets of code.
